markdownlint.js

```diff
@@ -131,30 +131,47 @@ function prepareFileList(files, fileExtensions, previousResults) {
 }
 
 function printResult(lintResult) {
+  let lintResultString = '';
   const results = flatten(Object.keys(lintResult).map(file => {
     return lintResult[file].map(result => {
-      return {
-        file: file,
-        lineNumber: result.lineNumber,
-        column: (result.errorRange && result.errorRange[0]) || 0,
-        names: result.ruleNames.join('/'),
-        description: result.ruleDescription +
-          (result.errorDetail ? ' [' + result.errorDetail + ']' : '') +
-          (result.errorContext ? ' [Context: "' + result.errorContext + '"]' : '')
-      };
+      if (options.json) {
+        return {
+          fileName: file,
+          ...result          
+        }        
+      } else {
+        return {
+          file: file,
+          lineNumber: result.lineNumber,
+          column: (result.errorRange && result.errorRange[0]) || 0,
+          names: result.ruleNames.join('/'),
+          description: result.ruleDescription +
+            (result.errorDetail ? ' [' + result.errorDetail + ']' : '') +
+            (result.errorContext ? ' [Context: "' + result.errorContext + '"]' : '')
+        };
+      }
     });
   }));
-  let lintResultString = '';
+  
   if (results.length > 0) {
-    results.sort((a, b) => {
-      return a.file.localeCompare(b.file) || a.lineNumber - b.lineNumber ||
-        a.names.localeCompare(b.names) || a.description.localeCompare(b.description);
-    });
-    lintResultString = results.map(result => {
-      const {file, lineNumber, column, names, description} = result;
-      const columnText = column ? `:${column}` : '';
-      return `${file}:${lineNumber}${columnText} ${names} ${description}`;
-    }).join('\n');
+    if (options.json) {
+      results.sort((a, b) => {
```

Comment:
I wonder if we could reuse the same call to `sort` here? I think you're using `fileName` for consistency with CLI2? Then what about changing the existing `file` to `fileName` since it's internal only? Then if we did something like `((a.names && b.names) ? a.names.localeCompare(b.names) : 0)` and reuse the call to `sort`?

Reply:
Put differently: call `results.sort`, then `if (options.json) { stringify } else { map }`

Reply:
Yeah - I wanted to write it that way, but the names and ruleNames are not the same structure form. One is collapsed with a join, the other is a list. So while fileName is totally reusable for the sort, the other tie-breakers - names, description are not stable. Maybe it doesn't matter. Happy to use just the fileName, lineNumber (past that point I am not sure it matters). 

Your project, your call.